src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -35,27 +35,25 @@ JvmtiTagMapKey::JvmtiTagMapKey(oop obj) : _obj(obj) {}
 
 JvmtiTagMapKey::JvmtiTagMapKey(const JvmtiTagMapKey& src) {
   // move object into WeakHandle when copying into the table
-  assert(src._obj != nullptr, "must be set");
+  if (src._obj != nullptr) {
 
-  // obj was read with AS_NO_KEEPALIVE, or equivalent, like during
-  // a heap walk.  The object needs to be kept alive when it is published.
-  Universe::heap()->keep_alive(src._obj);
+    // obj was read with AS_NO_KEEPALIVE, or equivalent, like during
+    // a heap walk.  The object needs to be kept alive when it is published.
+    Universe::heap()->keep_alive(src._obj);
 
-  _wh = WeakHandle(JvmtiExport::weak_tag_storage(), src._obj);
-  _obj = nullptr;
-}
-
-JvmtiTagMapKey::~JvmtiTagMapKey() {
-  // If obj is set null it out, this is called for stack object on lookup,
-  // and it should not have a WeakHandle created for it yet.
-  if (_obj != nullptr) {
+    _wh = WeakHandle(JvmtiExport::weak_tag_storage(), src._obj);
     _obj = nullptr;
-    assert(_wh.is_null(), "WeakHandle should be null");
   } else {
-    _wh.release(JvmtiExport::weak_tag_storage());
+    // resizing needs to create a copy.
+    _wh = src._wh;
+    _obj = nullptr;
   }
```

Comment:
There can be just one line at 51 instead of two lines at 45 and 49.
Then, I do not see where in the class `JvmtiTagMapKey` the `_obj` can obtain non nullptr value.

Reply:
Yes, _obj should always be null after copying either because we've transferred the oop over to the WeakHandle or that it's a copy so already a WeakHandle.  I admit that having _obj assigned to null in both places looks odd, but it was for two different reasons.  I added a comment too.